pkg/expr/graph.go

```diff
@@ -55,7 +55,45 @@ type DataPipeline []Node
 // map of the refId of the of each command
 func (dp *DataPipeline) execute(c context.Context, now time.Time, s *Service) (mathexp.Vars, error) {
 	vars := make(mathexp.Vars)
+
+	// Execute datasource nodes first, and grouped by datasource.
```

Comment:
The parts related to doing the datasource first come from another PR that this one based of off, mentioned in the PR original issue notes.